design-document.md

```diff
@@ -14,4 +24,171 @@
+finite field. As such, the main ingredient of Prio is an implementation of
+arithmetic in a finite field suitable for the given application.
+
+We will use a prime field. The choice of prime is influenced by the following
+criteria:
+
+1. **Field size.** How big the field needs to be depends on the type of data
+   being aggregated and how many users there are. The field size also impacts
+   the security level: the longer the validity circuit, the larger the field
+   needs to be in order to affectively detect malicious clients.
+1. **Fast polynomial operations.** In order to make Prio practical, it's
+   important that implementations employ FFT to speed up polynomial operations.
+   In particular, the prime modulus p should be chosen so that (p-1) = 2^n * s
+   for large n and odd s. Then g^s is a principle, 2^n-th root of unity (i.e.,
+   g^(s\*2^n) = 1), where g is the generator of the multiplicative subgroup.
+   This fact allows us to quickly evaluate and interpolate polynomials at 2^d-th
+   roots of unity for 1 <= d <= n.
+1. **Highly composite subgroup.** Suppose that (p-1) = 2^n * s. It's best if s
+   is highly composite because this minimizes the number of multiplications
+   required to compute the inverse or apply Fermat's Little Theorem. (See
+   [BBC+19, Section 5.2].)
+
+The table below lists parameters that meet these criteria at various levels of
+security. (Note that \#2 is the field used in "Prio v2".) The "bits" column
+indicates the size of the prime. For example, for p=3221225473,
+bits=log2(3221225473), which is approximately 31.53.
+
+| #  | bits   | p                                      | g  | n   | s                |
+|----|--------|----------------------------------------|----|-----|------------------|
+| 1  | 31.58  | 3221225473                             | 5  | 30  | 3                |
+| 2  | 31.99  | 4293918721                             | 19 | 20  | 3^2 * 5 * 7 * 13 |
+| 3  | 63.75  | 15564440312192434177                   | 5  | 59  | 3^3              |
+| 4  | 66.17  | 83010348331692982273                   | 11 | 63  | 3^2              |
+| 5  | 122.81 | 9304595970494411110326649421962412033  | 3  | 120 | 7                |
+| 6  | 125.40 | 56076806072176138388129360355577036801 | 13 | 116 | 3^3 * 5^2        |
+
+#### Key encapsulation
+
+Our instantiation of the core protocol involves two additional operations:
+public key encryption and pseudorandom number generation. The combination of
+these primitives that we use here allows us to make an additional
+simplification. If we assume that a secure channel will be used between the user
+and the leader (e.g., a server-authenticated TLS connection), then there is no
+need to encrypt the leaders input/proof share. In fact, all that we need to
+encrypt are seeds for a PRG. As a result, in our application we can get away
+with using a simple key-encapsulation mechanism (KEM) rather than full-blown
+encryption.
```

Comment:
```suggestion
these primitives that we use here allows us to make an additional
simplification. However, we assume users and leaders communicate over a secure,
authenticated channel, such as TLS. As a result, we only need to encrypt
CSPRNG seeds, which requires only a key-encapsulation mechanism (KEM) 
rather than full-blown encryption.
```